File: nitro-fs/package/android/src/main/java/com/margelo/nitro/nitrofs/HybridFs.kt
package com.margelo.nitro.nitrofs

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.io.File


class HybridFs : HybridFsSpec() {
    override val documentsDirectory: String
        get() = context.filesDir.absolutePath
    override val cacheDirectory: String
        get() = context.cacheDir.absolutePath
    override val tempDirectory: String
        get() = context.cacheDir.absolutePath + "/tmp" // Android doesn't have a separate temp directory, using cache directory instead, and appending /tmp to avoid conflicts with cache files

    override fun readFile(path: String): Promise<String> {
        return Promise.async {
            File(path).readText(Charsets.UTF_8)
        }
    }

    override fun writeFile(
        path: String, content: String
    ): Promise<Unit> {
        return Promise.async {
            File(path).writeText(content, Charsets.UTF_8)
            Unit
        }
    }

    override fun deleteFile(path: String): Promise<Unit> {
        return Promise.async {
            File(path).delete()
            Unit
        }
    }

    override fun mkdir(path: String): Promise<Unit> {
        return Promise.async {
            File(path).mkdirs()
            Unit
        }
    }

    override fun readDir(path: String): Promise<Array<FileEntry>> {
        return Promise.async {
            File(path).listFiles()?.map { file ->
                FileEntry(
                    name = file.name,
                    path = file.absolutePath,
                    isDirectory = file.isDirectory,
                    size = file.length().toDouble(),
                    lastModified = file.lastModified().toDouble()
                )
            }?.toTypedArray() ?: emptyArray()
        }
    }

    override fun exists(path: String): Promise<Boolean> {
        return Promise.async {
            File(path).exists()
        }
    }

    override fun move(
        srcPath: String, destPath: String
    ): Promise<Unit> {
        return Promise.async {
            File(srcPath).renameTo(File(destPath))
            Unit
        }
    }

    override fun copy(
        srcPath: String, destPath: String
    ): Promise<Unit> {
        return Promise.async {
            File(srcPath).copyTo(File(destPath))
            Unit
        }
    }


    companion object {
        private val context =
            NitroModules.applicationContext ?: error("React native context not found")
    }

}

File: nitro-fs/package/ios/HybridFs.swift
import Foundation
import NitroModules

class HybridFs: HybridFsSpec {
    var documentsDirectory: String
    
    var cacheDirectory: String
    
    var tempDirectory: String
    
    override init() {
        documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first?.path ?? ""
        cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path ?? ""
        tempDirectory = NSTemporaryDirectory()
    }
    
    func readFile(path: String) throws -> NitroModules.Promise<String> {
        Promise.async {
            return try String(contentsOfFile: path, encoding: .utf8)
        }
    }
    
    func writeFile(path: String, content: String) throws -> NitroModules.Promise<Void> {
        Promise.async {
            try content.write(toFile: path, atomically: true, encoding: .utf8)
        }
    }
    
    func deleteFile(path: String) throws -> NitroModules.Promise<Void> {
        Promise.async {
            try FileManager.default.removeItem(atPath: path)
        }
    }
    
    func mkdir(path: String) throws -> NitroModules.Promise<Void> {
        Promise.async {
            try FileManager.default.createDirectory(atPath: path, withIntermediateDirectories: true, attributes: nil)
        }
    }
    
    func readDir(path: String) throws -> NitroModules.Promise<[FileEntry]> {
        Promise.async {
            let contents = try FileManager.default.contentsOfDirectory(atPath: path)
            return contents.map { name in 
                let fullPath = (path as NSString).appendingPathComponent(name)
                let attributes = try? FileManager.default.attributesOfItem(atPath: fullPath)
                let isDirectory = (attributes?[.type] as? FileAttributeType) == .typeDirectory
                let size = attributes?[.size] as? UInt64 ?? 0
                let lastModified = (attributes?[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0

                return FileEntry(name: name, path: fullPath, isDirectory: isDirectory, size: Double(size), lastModified: lastModified)
            }
            
        }
    }
    
    func exists(path: String) throws -> NitroModules.Promise<Bool> {
        Promise.async {
            return FileManager.default.fileExists(atPath: path)
        }
    }
    
    func move(srcPath: String, destPath: String) throws -> NitroModules.Promise<Void> {
        Promise.async {
            try FileManager.default.moveItem(atPath: srcPath, toPath: destPath)
        }
    }
    
    func copy(srcPath: String, destPath: String) throws -> NitroModules.Promise<Void> {
        Promise.async {
            try FileManager.default.copyItem(atPath: srcPath, toPath: destPath)
        }
    }
    
    
}
